app/components/navbar.tsx

```diff
@@ -65,45 +66,76 @@ function NavLink({
 
 const iconTransformOrigin = {transformOrigin: '50% 100px'}
 function DarkModeToggle({variant = 'icon'}: {variant?: 'icon' | 'labelled'}) {
-  const [, setTheme] = useTheme()
+  const requestInfo = useRequestInfo()
+  const fetcher = useFetcher()
+
+  const optimisticMode = useOptimisticThemeMode()
+  const mode = optimisticMode ?? requestInfo.userPrefs.theme ?? 'system'
+  const nextMode =
+    mode === 'system' ? 'light' : mode === 'light' ? 'dark' : 'system'
+
   return (
-    <button
-      onClick={() => {
-        setTheme(previousTheme =>
-          previousTheme === Theme.DARK ? Theme.LIGHT : Theme.DARK,
-        )
-      }}
-      className={clsx(
-        'border-secondary hover:border-primary focus:border-primary inline-flex h-14 items-center justify-center overflow-hidden rounded-full border-2 p-1 transition focus:outline-none',
-        {
-          'w-14': variant === 'icon',
-          'px-8': variant === 'labelled',
-        },
-      )}
-    >
-      {/* note that the duration is longer then the one on body, controlling the bg-color */}
-      <div className="relative h-8 w-8">
-        <span
-          className="absolute inset-0 rotate-90 transform text-black transition duration-1000 motion-reduce:duration-[0s] dark:rotate-0 dark:text-white"
-          style={iconTransformOrigin}
-        >
-          <MoonIcon />
-        </span>
+    <fetcher.Form method="POST" action="action/set-theme">
```

Comment:
```suggestion
    <fetcher.Form method="POST" action="/action/set-theme">
```

Otherwise this will be relative to wherever the component is rendered.